scipy/signal/_signaltools.py

```diff
@@ -1919,16 +1920,95 @@ def correlate2d(in1, in2, mode='full', boundary='fill', fillvalue=0):
     if swapped_inputs:
         in1, in2 = in2, in1
 
-    val = _valfrommode(mode)
-    bval = _bvalfromboundary(boundary)
-    out = _sigtools._convolve2d(in1, in2.conj(), 0, val, bval, fillvalue)
+    use_fft = _correlate2d_fft_compatible(in1, in2, mode, boundary, fillvalue)
+
+    if use_fft:
+        out = _correlate2d_fft_fill(in1, in2, mode)
+    else:
+        kernel = np.conjugate(in2[::-1, ::-1])
+        out = convolve2d(in1, kernel, mode=mode, boundary=boundary,
+                         fillvalue=fillvalue)
 
     if swapped_inputs:
         out = out[::-1, ::-1]
 
     return xp.asarray(out)
 
 
+def _correlate2d_fft_compatible(in1, in2, mode, boundary, fillvalue):
+    """Return True when the FFT implementation can be used."""
+    # Ensure the mode string is valid by reusing the shared checker.
+    _ = _valfrommode(mode)
+
+    boundary = boundary.lower() if isinstance(boundary, str) else boundary
+    if boundary not in {"fill", "pad"}:
+        return False
+
+    # Only zero padding is currently supported for the FFT implementation;
+    # other fill values require additional adjustments that are not covered here.
+    if fillvalue != 0:
+        return False
```

Comment:
not currently hit by tests based on local testing via `spin test -j 12 --coverage`

<img width="695" height="94" alt="image" src="https://github.com/user-attachments/assets/fcae40be-3e8d-42fc-8b5e-b4a1dd20bebf" />